zhusuan/distributions/special.py

```diff
@@ -0,0 +1,144 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+from __future__ import absolute_import
+from __future__ import division
+
+import tensorflow as tf
+from numpy import inf
+
+from zhusuan.distributions.base import Distribution
+from zhusuan.distributions.utils import assert_same_float_and_int_dtype, \
+    maybe_explicit_broadcast
+
+
+__all__ = [
+    'Empirical',
+    'Implicit',
+]
+
+
+class Empirical(Distribution):
+    """
+    The class of Empirical distribution.
+    See :class:`~zhusuan.distributions.base.Empirical` for details.
+
+    :param shape: A list or tuple describing the `batch_shape` of the distribution.
+        The entries of the list can either be int, Dimension or Tensor.
+    :param dtype: The value type of samples from the distribution.
+    :param group_ndims: A 0-D `int32` Tensor representing the number of
+        dimensions in `batch_shape` (counted from the end) that are grouped
+        into a single event, so that their probabilities are calculated
+        together. Default is 0, which means a single value is an event.
+        See :class:`~zhusuan.distributions.base.Distribution` for more detailed
+        explanation.
+    :param is_continuous: Whether the distribution is continous or not.
+        If None will consider it continuous only if `dtype` is a float type.
+    """
+
+    def __init__(self, shape, dtype,
+                 group_ndims=0,
+                 is_continuous=None,
+                 **kwargs):
+        try:
+            self.static_shape = tf.TensorShape(shape[0])
+        except Exception:
+            self.static_shape = tf.TensorShape(tf.Dimension(None))
+        for s in shape[1:]:
+            try:
+                self.static_shape = self.static_shape.concatenate(tf.TensorShape(s))
+            except Exception:
+                self.static_shape = self.static_shape.concatenate(tf.Dimension(None))
```

Comment:
I guess there should be some internal TF implementation of this feature. From list or TensorShape to TensorShape.